File: 2015/day03_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Position = tuple[int, int]


def solution(content: str) -> Any:
    santa: Position = (0, 0)
    robo: Position = (0, 0)
    visited: list[Position] = [santa, robo]
    for direction in content[::2]:
        santa = move(santa, direction)
        visited.append(santa)
    for direction in content[1::2]:
        robo = move(robo, direction)
        visited.append(robo)
    return len(
        [True for index, item in enumerate(visited) if item not in visited[:index]]
    )


def move(current: Position, direction: str) -> Position:
    if direction == "^":
        return (current[0], current[1] + 1)
    if direction == "v":
        return (current[0], current[1] - 1)
    if direction == "<":
        return (current[0] - 1, current[1])
    if direction == ">":
        return (current[0] + 1, current[1])
    raise ValueError(f"Invalid direction: {direction}")


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2015/day03-2.ts
type Position = {
  x: number;
  y: number;
};

function solution(content: string) {
  return content
    .split("")
    .reduce(
      (acc, direction, index) =>
        index % 2 === 0 ? [...acc, move(acc.at(-1)!, direction)] : acc,
      [{ x: 0, y: 0 }]
    )
    .concat(
      content
        .split("")
        .reduce(
          (acc, direction, index) =>
            index % 2 !== 0 ? [...acc, move(acc.at(-1)!, direction)] : acc,
          [{ x: 0, y: 0 }]
        )
    )
    .filter(
      (item, index, array) =>
        !array
          .slice(0, index)
          .find((itm) => itm.x === item.x && itm.y === item.y)
    ).length;
}

function move({ x, y }: Position, direction: string) {
  if (direction === "^") return { x, y: y + 1 };
  if (direction === "v") return { x, y: y - 1 };
  if (direction === "<") return { x: x - 1, y };
  if (direction === ">") return { x: x + 1, y };
  throw new Error(`Invalid direction ${direction}`);
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
